File: platforms/android/java/app/msime/client/LetterKeyFacePolicy.java
package app.msime.client;

import java.util.Locale;

/** Keeps the visible letter face separate from the case sent to the input engine. */
public final class LetterKeyFacePolicy {
    private LetterKeyFacePolicy() { }

    public static boolean displaysUppercase(boolean chineseMode, boolean localMode,
            boolean shifted) {
        return (chineseMode && !localMode) || (!chineseMode && shifted);
    }

    public static String face(String lowercase, boolean chineseMode, boolean localMode,
            boolean shifted) {
        if (lowercase == null || lowercase.isEmpty()) return "";
        return displaysUppercase(chineseMode, localMode, shifted)
            ? lowercase.toUpperCase(Locale.ROOT) : lowercase;
    }

    public static String accessibilityLabel(String lowercase, boolean chineseMode,
            boolean localMode, boolean shifted) {
        if (lowercase == null || lowercase.isEmpty()) return "字母";
        String uppercase = lowercase.toUpperCase(Locale.ROOT);
        return (!chineseMode && shifted ? "大写 " : "字母 ") + uppercase;
    }
}

File: platforms/harmony/entry/src/main/ets/keyboard/input/LetterKeyFacePolicy.ts
/**
 * Keeps the visible letter face separate from the case sent to the Engine, ported from
 * platforms/android/java/app/msime/client/LetterKeyFacePolicy.java.
 *
 * Chinese mode prints uppercase faces while still sending lowercase, which is why the face and the
 * engine input are decided separately.
 */
export class LetterKeyFacePolicy {
  static displaysUppercase(chineseMode: boolean, localMode: boolean, shifted: boolean): boolean {
    return (chineseMode && !localMode) || (!chineseMode && shifted);
  }

  static face(lowercase: string | null, chineseMode: boolean, localMode: boolean,
              shifted: boolean): string {
    if (lowercase === null || lowercase.length === 0) {
      return '';
    }
    return LetterKeyFacePolicy.displaysUppercase(chineseMode, localMode, shifted)
      ? lowercase.toUpperCase() : lowercase;
  }

  static accessibilityLabel(lowercase: string | null, chineseMode: boolean, localMode: boolean,
                            shifted: boolean): string {
    if (lowercase === null || lowercase.length === 0) {
      return '字母';
    }
    return (!chineseMode && shifted ? '大写 ' : '字母 ') + lowercase.toUpperCase();
  }
}
